File: Sliding Window/Minimum Swaps to Group All 1's Together II/solution.py
class Solution:
    def minSwaps(self, nums: List[int]) -> int:
        n = len(nums)
        ones = 0
        for num in nums:
            if num == 1:
                ones += 1
        result = n    
        start = 0
        length = 0
        for i in range(ones):
            if nums[i] == 1:
                length += 1
        result = min(result, ones - length)
        
        for i in range(ones, 2*n):
            i = i%n
            if nums[start] == 1:
                length -= 1
            if nums[i] == 1:
                length += 1
            start += 1
            start %= n
            result = min(result, ones - length)
        return result

File: Sliding Window/Minimum Swaps to Group All 1's Together II/solution.cpp
class Solution {
public:
    int minSwaps(vector<int>& nums) {
        int n = nums.size();
        int ones = count(nums.begin(), nums.end(), 1);
        int running_ones = 0;
        int result = ones;
        for(int i = 0; i < n + ones; i++) {
            if(i >= ones) {
                running_ones -= nums[(i-ones+n)%n] == 1;
            }
            running_ones += nums[i%n] == 1;
            if(i >= ones) {
                result = min(result, ones - running_ones);
            }
        }
        return result;
    }
};
